### PR TITLE
[Feature] Support to html pasting on web

### LINKED ISSUE #3790: [HOLD for payment July 20] Markdown - Copied text (with markdown) does not show the formatting when pasted in e.cash
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!

Upwork job: https://www.upwork.com/jobs/~01784929033238eacd
___

## Action Performed:
1. Copy the following text: 

**Bold text**
_italic text_
~Strikethrough text~

- Bullet points test 
- Another Bullet point

  Indents test 
2. Paste the text in e.cash

## Expected Result:
Formatting should be displayed in e.cash. 

## Actual Result:
Text is pasted without any formatting. Only line breaks are displayed. 

## Workaround:
User has to edit the text in e.cash compose box and add the markdown manually. 

## Platform:
Where is this issue occurring?

 Web ✔️ 
 iOS
 Android
 Desktop App
 Mobile Web

**Version Number:** 1.0.74-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

image

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___
From @mallenexpensify https://expensify.slack.com/archives/C01GTK53T8Q/p1617291564288600
>When copying text (gmail draft in this example) and pasting into e.cash, formatting is being stripped. I can create an issue for it, wondering if we should attempt to fix parts of it first, like line breaks, then markdown in a separate issue? Thoughts/ideas?

### PR DESCRIPTION
**Pull request by parasharrajat**
### Fixed Issues

$ Fixes #3790

### Tests | QA Steps

 1. Copy the following text:

 **Bold text**
_italic text_
~Strikethrough text~

 2. Paste the text in e.cash

### Tested On

- [x] Web
- [ ] Mobile Web
- [x] Desktop
- [ ] iOS
- [ ] Android

#### Web | Desktop

https://user-images.githubusercontent.com/24370807/125504119-ceb4a52b-fabd-4e21-aca6-d070a1f76637.mp4

### CODE REVIEW

**Review by Jag96 — CHANGES_REQUESTED** [review]
A couple of nitpicks but it's working well!

File: src/components/TextInputFocusable/index.js [hunk 1]
```diff
@@ -241,25 +244,50 @@ class TextInputFocusable extends React.Component {
...
+     * Otherwise, convert pasted HTML to Markdown and set it on the composer.
      *
      * @param {ClipboardEvent} event
      */
-    checkForAttachment(event) {
+    handlePaste(event) {
         const {files, types} = event.clipboardData;
         const TEXT_HTML = 'text/html';
-        const TEXT_PLAIN = 'text/plain';
+        const pastedHTML = event.clipboardData.getData(TEXT_HTML);
```
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/components/TextInputFocusable/index.js:273]: Yes. Thanks

File: src/components/TextInputFocusable/index.js [hunk 2]
```diff
@@ -241,25 +244,50 @@ class TextInputFocusable extends React.Component {
...
+            return;
+        }
+
+        // If paste contains HTML
+        if (types.includes(TEXT_HTML)) {
+            event.preventDefault();
             const domparser = new DOMParser();
-            const embededImages = domparser.parseFromString(event.clipboardData.getData(TEXT_HTML), TEXT_HTML).images;
-            const pastedText = event.clipboardData.getData(TEXT_PLAIN);
+            const embededImages = domparser.parseFromString(pastedHTML, TEXT_HTML).images;
```
[the remark this answers is not in the document]
Comment (parasharrajat) [reply src/components/TextInputFocusable/index.js:287]: Yup. It was already there but I will update.

**Comment by parasharrajat** [comment]
@Jag96 While handling all this I found a couple of issues.
1. If we edited the formatted text I see MD syntax but clicking save does not show formatting anymore.
image
2. There is an extra Line Break added on paste. I debugged the main cause and found we are getting newlines via HTMLparser.
3. When I copy the existing sent message which is formatted and pastes it. it does not convert to formatted MD strings. The main issue here is that we don't use HTML semantic tags for rendering Every type of element is made with div and styling rules. Which we can't really interpret ATM.

**Comment by parasharrajat** [comment]
Updated.

**Comment by Jag96** [comment]
Good finds!

- `1` sounds like a regression to me, so I think that would be a blocker in this case and something we should look into.
- For `2`, I tested on Firefox and Chrome and it looks like Firefox has extra line breaks in the HTML whereas Chrome doesn't. Could that be the issue there? If our code is adding unnecessary line breaks I think that'd be a blocker, but if it's just the way the HTML is copied from the browser I don't think that should block this PR. Looking into this a bit more, I do see that there are `<br>` tags in the HTML when pasting on chrome (using the same text in the [issue description](https://github.com/Expensify/Expensify.cash/issues/3790)), so perhaps there is an issue w/ the newline regex causing it to fail in this case?
- For `3` that sounds like a different issue and something to discuss in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) since I think that will be a bit more involved, so I wouldn't block this PR on that.

Thoughts?

**Comment by parasharrajat** [comment]
Ok. I agree. Should I add those fixes in this PR?

**Comment by Jag96** [comment]
I think fixes for 1 and 2 should be added to this PR, assuming 2 requires a fix.

**Comment by parasharrajat** [comment]
@Jag96 Just opened a new PR for `2`. https://github.com/Expensify/expensify-common/pull/393. Could be merged quickly and we can move forward here.

**Comment by parasharrajat** [comment]
Okay, I have updated the E-common here. and `1` & `2` is fixed

**Comment by parasharrajat** [comment]
Please let me know if anything else is needed here.

**Comment by Jag96** [comment]
1 looks good, for 2 it looks like when I copy HTML that contains a `<br>` tag it doesn't add the line breaks if the `<br>` tag has styles. For example, copying the following from https://github.com/Expensify/Expensify.cash/issues/3790:

image

Results in the following `html` value inside `handlePastedHTML`
```
"<meta charset='utf-8'><strong style=\"box-sizing: border-box; font-weight: 600; color: rgb(36, 41, 46); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;\">Bold text</strong><br style=\"box-sizing: border-box; color: rgb(36, 41, 46); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;\"><em style=\"box-sizing: border-box; color: rgb(36, 41, 46); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;\">italic text</em><br style=\"box-sizing: border-box; color: rgb(36, 41, 46); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;\"><del style=\"box-sizing: border-box; color: rgb(36, 41, 46); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255);\">Strikethrough text</del>"
```

Before this PR, pasting the content would include these line breaks, but now it strips them:

| Before | After |
|---|---|
| image | image |

However, this just looks like a bug in the regex and isn't really tied to the core of this PR, so I won't block on it and I'll open a new issue for this.

**Review by Jag96 — APPROVED** [review]

**Comment by parasharrajat** [comment]
Yeah @Jag96  Do you want me to correct the Br tag regex

**Comment by Jag96** [comment]
@parasharrajat sounds good, feel free to leave a comment on https://github.com/Expensify/Expensify.cash/issues/4052 so I can assign you, I'll invite you to the Upwork job in the meantime

**Comment by isagoico** [comment]
@Jag96 Since this PR is failing because of the [issue](https://github.com/Expensify/Expensify.cash/issues/4052) above, should we block prod deploy on this? or can we check it off the PR list and test this again when the fix is merged?

**Comment by Jag96** [comment]
No worries! Now that https://github.com/Expensify/Expensify.cash/pull/4067 is on staging that'll be the latest fix

**Comment by sonialiap** [comment]
An error was identified that was linked to this PR - https://github.com/Expensify/App/issues/9505

**Pull request merged**